### DISCUSSION TITLE
Get-EnvironmentalVariable

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [SamErde/PSPreworkout#23] [General]
I like the idea of a command that makes it easier to view environmental data. However, I think you can improve this by defaulting to the `Process` scope.  A user is likely to use the command like this after seeing the full output.

```powershell
Get-EnvironmentVariable -Name username
```

I expected to see `Jeff`. Then I remembered environment scope.

```powershell
PS C:\> Get-EnvironmentVariable -Name USERNAME -Target User
PS C:\> Get-EnvironmentVariable -Name USERNAME -Target Machine
SYSTEM
PS C:\> Get-EnvironmentVariable -Name USERNAME -Target PROCESS
Jeff
```

Your default output shows the Process `Username` so I would expect that to be the value when I get a single environment variable.

And again, instead of writing a hashtable as the output, write an object. That would allow the user to run `Get-EnvironmentVariable | Sort-Object`.  It would also be helpful to be able to use wildcards: `Get-EnvironmentVariable w*`.

Since I know what the output object is, I can do it:

```powershell
PS C:\> (Get-EnvironmentVariable).GetEnumerator() | Where {$_.key -match "^w"}

Name                           Value
----                           -----
WT_SESSION                     afbe1a21-4a72-411f-9cb3-e6c95d495fd3
windir                         C:\WINDOWS
WT_PROFILE_ID                  {0b9ec627-34ae-45ba-8a93-9be14f576b34}
WSLENV                         WT_SESSION:WT_PROFILE_ID:
```

But the whole point of the command is to reduce friction for the user.

Finally, if you are going to write a custom object to the pipeline, I would try to show what target scopes apply to each environment variable.

**Comment by SamErde** [comment]
Great insights. I'm slowly starting to wrap my head around all of places that you've pointed out these principles and over time it is having / will have an increasing impact on how I approach every new ( or refactored) function. Writing these gev and sev functions already provided some new things learned and this layers nicely on that.

**Comment by SamErde** [comment]
I just wrapped up the object output and will be testing ASAP. Still have a bit to do for partial name matches and other basic logic, but I'm happy with the basic output now. 
https://github.com/SamErde/PSPreworkout/blob/main/src/PSPreworkout/Public/Get-EnvironmentVariable.ps1
